src/v/cloud_storage/access_time_tracker.cc

```diff
@@ -54,21 +45,140 @@ void read_nested(
 
 namespace cloud_storage {
 
+// For async serialization/deserialization, the fixed-size content
+// is defined in a header struct, and then the main body of the encoding
+// is defined by hand in the read/write methods so that it can be done
+// with streaming.
+struct table_header
+  : serde::envelope<table_header, serde::version<0>, serde::compat_version<0>> {
+    size_t table_size{0};
+
+    auto serde_fields() { return std::tie(table_size); }
+};
+
+ss::future<> access_time_tracker::write(ss::output_stream<char>& out) {
+    // This lock protects us from the _table being mutated while we
+    // are iterating over it and yielding during the loop.
+    auto lock_guard = co_await ss::get_units(_table_lock, 1);
+
+    _dirty = false;
+
+    const table_header h{.table_size = _table.size()};
+    iobuf header_buf;
+    serde::write(header_buf, h);
+    co_await write_iobuf_to_output_stream(std::move(header_buf), out);
+
+    // How many items to serialize per stream write()
+    constexpr size_t chunk_count = 2048;
+
+    size_t i = 0;
+    iobuf serialize_buf;
+    for (auto it : _table) {
+        serde::write(serialize_buf, it.first);
+        serde::write(serialize_buf, it.second);
+        ++i;
+        if (i % chunk_count == 0 || i == _table.size()) {
+            for (const auto& f : serialize_buf) {
+                co_await out.write(f.get(), f.size());
+            }
+            serialize_buf.clear();
+        }
+    }
+
+    lock_guard.return_all();
+    on_released_table_lock();
+}
+
+void access_time_tracker::on_released_table_lock() {
+    // Before dropping lock, drain any pending upserts that came in via
+    // add/remove calls while we were locking the main table.
+    for (const auto& [hash, ts] : _pending_upserts) {
```

Comment:
Should we be marking `_dirty = true` here? If the table was locked during `add_timestamp` we omit marking it dirty (presumably because that should only happen when `_table` is successfully updated). Or is the idea to only mark dirty if there's been an add since the last time we serialized (rather than really trying to track if there are changes that haven't been serialized, since this tracking is best-effort)

Reply:
I think it's fair to set dirty=true here -- it's not very important because this map is always handled in a kinda sloppy way, but makes the code easier to reason about.  I've added that.